Jagged Arrays

Is it impossible to create jagged arrays? Trying to compile the following code:
```CSharp
        public static int JaggedArray(int a, int b)
        {

            int[] array1 = new int[] { 1, 2, 3, 4 };
            int[] array2 = new int[] { 5, 6, 7, 8 };
            int[] array3 = new int[] { 1, 3, 2, 1 };
            int[] array4 = new int[] { 5, 4, 3, 2 };

            int[][] arrayArray = new int[][] { array1, array2, array3, array4 };

            return arrayArray[a][b];
        }
```
results in:
```console
Microsoft(R) Build Engine versão 16.11.0+0538acc04 para .NET
Copyright (C) Microsoft Corporation. Todos os direitos reservados.

  Determinando os projetos a serem restaurados...
  Todos os projetos estão atualizados para restauração.
  TestArrayContract -> C:\Users\kenji\OneDrive\Documentos\Workspace\debugger\contracts\TestArray\src\bin\Debug\net5.0\TestArrayContract.dll
    Determinando os projetos a serem restaurados...
    Todos os projetos estão atualizados para restauração.
C:\Users\kenji\OneDrive\Documentos\Workspace\debugger\contracts\TestArray\src\TestArrayContract.cs(46,38): error NC2008: Unsupported array rank: [][] [C:\Users\kenji\OneDrive\Documentos\Workspace\debugger\contracts\TestArray\src\TestArrayContract.csproj]
  Compilation failed.
C:\Users\kenji\OneDrive\Documentos\Workspace\debugger\contracts\TestArray\src\TestArrayContract.csproj(21,5): error MSB3073: O comando "dotnet nccs --debug --no-optimize "C:\Users\kenji\OneDrive\Documentos\Workspace\debugger\contracts\TestArray\src\TestArrayContract.csproj"" foi encerrado com o código 1.

FALHA da compilação.

C:\Users\kenji\OneDrive\Documentos\Workspace\debugger\contracts\TestArray\src\TestArrayContract.cs(46,38): error NC2008: Unsupported array rank: [][] [C:\Users\kenji\OneDrive\Documentos\Workspace\debugger\contracts\TestArray\src\TestArrayContract.csproj]
C:\Users\kenji\OneDrive\Documentos\Workspace\debugger\contracts\TestArray\src\TestArrayContract.csproj(21,5): error MSB3073: O comando "dotnet nccs --debug --no-optimize "C:\Users\kenji\OneDrive\Documentos\Workspace\debugger\contracts\TestArray\src\TestArrayContract.csproj"" foi encerrado com o código 1.
    0 Aviso(s)
    2 Erro(s)

Tempo Decorrido 00:00:05.40
```
Looking at the code at [MethodConvert](https://github.com/neo-project/neo-devpack-dotnet/blob/815c17f457783387da1cebbe6dcfbacd23961a7a/src/Neo.Compiler.CSharp/MethodConvert.cs) there are a lot of `if (numberOfArgumentsOrRanks != 1) {throw new CompilationException()}`. So is it by design to not permit the user to create jagged arrays?

--- a/src/Neo.Compiler.CSharp/MethodConvert.cs
+++ b/src/Neo.Compiler.CSharp/MethodConvert.cs
@@ -1244,8 +1244,6 @@ namespace Neo.Compiler
 
         private void ConvertArrayCreationExpression(SemanticModel model, ArrayCreationExpressionSyntax expression)
         {
-            if (expression.Type.RankSpecifiers.Count != 1)
-                throw new CompilationException(expression.Type, DiagnosticId.MultidimensionalArray, $"Unsupported array rank: {expression.Type.RankSpecifiers}");
             ArrayRankSpecifierSyntax specifier = expression.Type.RankSpecifiers[0];
             if (specifier.Rank != 1)
                 throw new CompilationException(specifier, DiagnosticId.MultidimensionalArray, $"Unsupported array rank: {specifier}");
@@ -1256,7 +1254,7 @@ namespace Neo.Compiler
                 if (type.ElementType.SpecialType == SpecialType.System_Byte)
                     AddInstruction(OpCode.NEWBUFFER);
                 else
-                    AddInstruction(new Instruction { OpCode = OpCode.NEWARRAY_T, Operand = new[] { (byte)type.GetStackItemType() } });
+                    AddInstruction(new Instruction { OpCode = OpCode.NEWARRAY_T, Operand = new[] { (byte)type.ElementType.GetStackItemType() } });
             }
             else
             {

--- a/tests/Neo.Compiler.CSharp.UnitTests/TestClasses/Contract_Array.cs
+++ b/tests/Neo.Compiler.CSharp.UnitTests/TestClasses/Contract_Array.cs
@@ -11,6 +11,15 @@ namespace Neo.Compiler.CSharp.UnitTests.TestClasses
 
     public class Contract_Array : SmartContract.Framework.SmartContract
     {
+        public static int[][] TestJaggedArray()
+        {
+            int[] array1 = new int[] { 1, 2, 3, 4 };
+            int[] array2 = new int[] { 5, 6, 7, 8 };
+            int[] array3 = new int[] { 1, 3, 2, 1 };
+            int[] array4 = new int[] { 5, 4, 3, 2 };
+            return new int[][] { array1, array2, array3, array4 };
+        }
+
         public static object TestIntArray()
         {
             var arrobj = new int[3];

--- a/tests/Neo.Compiler.CSharp.UnitTests/TestClasses/Contract_Array.cs
+++ b/tests/Neo.Compiler.CSharp.UnitTests/TestClasses/Contract_Array.cs
@@ -20,6 +20,15 @@ namespace Neo.Compiler.CSharp.UnitTests.TestClasses
             return new int[][] { array1, array2, array3, array4 };
         }
 
+        public static byte[][] TestJaggedByteArray()
+        {
+            byte[] array1 = new byte[] { 1, 2, 3, 4 };
+            byte[] array2 = new byte[] { 5, 6, 7, 8 };
+            byte[] array3 = new byte[] { 1, 3, 2, 1 };
+            byte[] array4 = new byte[] { 5, 4, 3, 2 };
+            return new byte[][] { array1, array2, array3, array4 };
+        }
+
         public static object TestIntArray()
         {
             var arrobj = new int[3];

--- a/tests/Neo.Compiler.CSharp.UnitTests/UnitTest_Array.cs
+++ b/tests/Neo.Compiler.CSharp.UnitTests/UnitTest_Array.cs
@@ -9,6 +9,19 @@ namespace Neo.Compiler.CSharp.UnitTests
     public class UnitTest_Array
     {
         [TestMethod]
+        public void Test_JaggedArray()
+        {
+            var testengine = new TestEngine();
+            testengine.AddEntryScript("./TestClasses/Contract_Array.cs");
+            var result = testengine.ExecuteTestCaseStandard("testJaggedArray");
+
+            var arr = result.Pop<Array>();
+            Assert.AreEqual(4, arr.Count);
+            var element0 = (Array)arr[0];
+            CollectionAssert.AreEqual(new int[] { 1, 2, 3, 4 }, element0.Cast<PrimitiveType>().Select(u => (int)u.GetInteger()).ToArray());
+        }
+
+        [TestMethod]
         public void Test_IntArray()
         {
             var testengine = new TestEngine();

--- a/tests/Neo.Compiler.CSharp.UnitTests/UnitTest_Array.cs
+++ b/tests/Neo.Compiler.CSharp.UnitTests/UnitTest_Array.cs
@@ -22,6 +22,19 @@ namespace Neo.Compiler.CSharp.UnitTests
         }
 
         [TestMethod]
+        public void Test_JaggedByteArray()
+        {
+            var testengine = new TestEngine();
+            testengine.AddEntryScript("./TestClasses/Contract_Array.cs");
+            var result = testengine.ExecuteTestCaseStandard("testJaggedByteArray");
+
+            var arr = result.Pop<Array>();
+            Assert.AreEqual(4, arr.Count);
+            var element0 = (Buffer)arr[0];
+            CollectionAssert.AreEqual(new byte[] { 1, 2, 3, 4 }, element0.InnerBuffer);
+        }
+
+        [TestMethod]
         public void Test_IntArray()
         {
             var testengine = new TestEngine();

